rfcs/build-system/02_standard_package_output_folders.md

```diff
@@ -0,0 +1,86 @@
+
+Once 13.2.0 becomes the minimum Node requirement, there is little reason to build CommonJS at all, and I anticipate we'll be removing that folder.
+
+### Pros and Cons
+
+#### Pros
+
+- Almost no difference from what we currently have - only the `build:commonjs-only` task needs to be modified to output to `lib-commonjs`.
+- Our partners won't really notice changes here at all, since this has been the convention we've used for a long time for nearly all the packages currently being consumed.
+- ESM is becoming the standard that all platforms will snap to, hence why `lib` and not `lib-esm`.
+- Changes in output won't change the folder structure
+- Unchanging folder structures mean that full builds are required less
+- End users can predict which folders contain what format
+- Scalable. If we need other formats in our JS packages, we can update add more formats.
+- `CommonJS` was the standard of the past. More and more, modern node libraries will be moving to ESM as Node 13.2.0+ now supports it. I anticipate this format to phase out over the next few years as AMD has.
+
+#### Cons
+
+- Potentially `esm` might be changed as the "standard" library format in the future, which would require us to adjust what `lib` means.
+
+### Alternatives
+
+Many libraries in OSS have kept `lib` representing CommonJS and have added `es` to the mix as the folder for containing esm. This is very likely because they've existed for a while and didn't want to change their existing folder structure. We could also consider this convention. React itself avoids `lib` and explicitly uses `cjs` to indicate the output format. (It also has no ESM flavor.)
+
+* `/cjs` - CommonJS
+* `/es` - esm
+* `/amd` - amd
+* `/dist` - statics and bundles
+
+#### Pros:
+
+* Follows a few more OSS library conventions (though there isn't consistency here)
+* Won't have to revisit if the JavaScript flavor of the month changes later
+
+#### Cons:
+
+* A bigger shift than the original proposal - will impact nearly all of our packages
+* Definitely more work to clean things up to this format
+
+### How do other major packages organize their output?
+
+"Standards" haven't really emerged other than `dist` being the output folder. There is a commonality of using `es` as the esm folder. Here's a small sampling:
+
+* @angular/core - Uses `bundles` for output which includes a UMD bundle for commonjs support, and `fesm2015` for esm. 
+* @apollo/client - No folders; `index.js` at root is esm while `index.cjs.js` is CommonJS.
+* @material-ui/core - Uses a root `index.js` rollup entry for commonjs and `es` folder for esm.
+* antd - Uses `es` for esm, `lib` for commonjs.
+* immer - Uses only `dist` and rolls up a separate `esm` bundle.
+* react - Commonjs is in `cjs`, there's also a `umd` folder for bundles.
+* redux - Uses `es` for esm, `lib` for commonjs.
+* styled-components - Only `dist` with `s-c.esm.js` and `s-c.cjs.js`. They also emit a react-native `s-c.native.js` flavor.
```

Comment:
regarding `dist/`, I definitely agree with "following" suggestion/change
> I briefly mentioned in on of [my PR's regarding better DX - TS path aliases](https://github.com/microsoft/fluentui/pull/17161/files), but it got lost within changes...

I'd recommend to have only one `dist` file in root of monorepo

```
/coverage
  /packages/
     /react-button
/dist
   /packages/
     /react-button
       | - index.umd.js
       | - index.module.js
       | - index.modern.js
       | - index.d.ts
       | - Readme.md
       | - License.md
       | - package.json  // normalized package.json - contains only important data (deps/peerDeps/license,etc)- no scripts/devDeps and other "baggage"
/packages
   /react-button
/apps
  /todo-app
```

**Pros:**
- source code is not being mismatched with generated code
- all generated assets live under one roof
- faster cleanup 
- ease of processing

**Cons:**
- might not work for v0,7,8 as they are not leveraging TS path aliases - additional work would be needed

Reply:
> **important thing to answer - is this RFC for both legacy and converged?**

Any nontrivial changes will have to be for converged only (the RFC needs to be updated to mention this). https://github.com/microsoft/fluentui/pull/17298#issuecomment-794691384

Reply:
@Hotell I think 1 dist folder with rollup bundles makes tons of sense for small libraries. With converged packages where they're more single purposed, for example. It is also a very good tool for avoiding deeply nested imports; we may be able to totally omit the build output completely. (There are a few scenarios where it's useful, like documentation, not sure.)

It breaks down in monolithic suite packages that re-export a lot of things, because of a variety of reasons. More to parse unnecessarily, more prone to side effect problems, bundle graph doesn't show enough detail.

So I could get behind this for the converged smaller-purpose packages.

Reply:
What if I proposed this:

Legacy: we keep the same. lib (esm), lib-commonjs (commonjs), amd (amd)

Converged:

```
dist/
   react-button.esm.js
   react-button.d.ts
   react-button.common.js
   react-button.umd.js
```  

@layershifter @ecraig12345 thoughts?